provisioner/ansible/provisioner.go

```diff
@@ -307,50 +302,185 @@ func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.C
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
-		defer os.Remove(tf.Name())
-
-		host := fmt.Sprintf("%s ansible_host=127.0.0.1 ansible_user=%s ansible_port=%d\n",
-			p.config.HostAlias, p.config.User, p.config.LocalPort)
-		if p.ansibleMajVersion < 2 {
-			host = fmt.Sprintf("%s ansible_ssh_host=127.0.0.1 ansible_ssh_user=%s ansible_ssh_port=%d\n",
-				p.config.HostAlias, p.config.User, p.config.LocalPort)
+		p.config.AnsibleEnvVars[i] = envVar
+	}
+	// Interpolate extra vars to check for generated values like password and port
+	for i, arg := range p.config.ExtraArguments {
+		arg, err := interpolate.Render(arg, &p.config.ctx)
+		if err != nil {
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
+		p.config.ExtraArguments[i] = arg
+	}
 
-		w := bufio.NewWriter(tf)
-		w.WriteString(host)
-		for _, group := range p.config.Groups {
-			fmt.Fprintf(w, "[%s]\n%s", group, host)
+	// Set up proxy if host IP is missing or communicator type is wrong.
+	if p.config.UseProxy.False() {
+		hostIP := generatedData["Host"].(string)
+		if hostIP == "" {
+			ui.Error("Warning: use_proxy is false, but instance does" +
+				" not have an IP address to give to Ansible. Falling back" +
+				" to use localhost proxy.")
+			p.config.UseProxy = config.TriTrue
 		}
-
-		for _, group := range p.config.EmptyGroups {
-			fmt.Fprintf(w, "[%s]\n", group)
+		connType := generatedData["ConnType"]
+		if connType != "ssh" && connType != "winrm" {
+			ui.Error("Warning: use_proxy is false, but communicator is " +
+				"neither ssh nor winrm, so without the proxy ansible will not" +
+				" function. Falling back to localhost proxy.")
+			p.config.UseProxy = config.TriTrue
 		}
+	}
 
-		if err := w.Flush(); err != nil {
-			tf.Close()
-			return fmt.Errorf("Error preparing inventory file: %s", err)
+	privKeyFile := ""
+	if !p.config.UseProxy.False() {
+		// We set up the proxy if useProxy is either true or unset.
+		pkf, err := p.setupAdapterFunc(ui, comm)
+		if err != nil {
+			return err
 		}
-		tf.Close()
-		p.config.InventoryFile = tf.Name()
+		// This is necessary to avoid accidentally redeclaring
+		// privKeyFile in the scope of this if statement.
+		privKeyFile = pkf
+
 		defer func() {
-			p.config.InventoryFile = ""
+			log.Print("shutting down the SSH proxy")
+			close(p.done)
+			p.adapter.Shutdown()
 		}()
+
+		go p.adapter.Serve()
+
+		// Remove the private key file
+		if len(privKeyFile) > 0 {
+			defer os.Remove(privKeyFile)
+		}
+	} else {
+		connType := generatedData["ConnType"].(string)
+		switch connType {
+		case "ssh":
+			ui.Message("Not using Proxy adapter for Ansible run:\n" +
+				"\tUsing ssh keys from Packer communicator...")
+			// In this situation, we need to make sure we have the
+			// private key we actually use to access the instance.
+			SSHPrivateKeyFile := generatedData["SSHPrivateKeyFile"].(string)
+			if SSHPrivateKeyFile != "" {
+				privKeyFile = SSHPrivateKeyFile
+			} else {
```

Comment:
Nit on removing the if/else 
```suggestion
			privKeyFile := generatedData["SSHPrivateKeyFile"].(string)
			if privKeyFile == "" {

```